packages/contract/src/contract-factory.ts

```diff
@@ -132,39 +132,35 @@ export default class ContractFactory {
    * @returns A promise that resolves to the deployed contract instance.
    */
   async deployContract(deployContractOptions: DeployContractOptions = {}) {
-    if (!this.account) {
-      throw new FuelError(ErrorCode.ACCOUNT_REQUIRED, 'Cannot deploy Contract without account.');
-    }
-
-    const { configurableConstants } = deployContractOptions;
-
-    if (configurableConstants) {
-      this.setConfigurableConstants(configurableConstants);
-    }
+    const { contractId, transactionRequest } = await this.prepareDeploy(deployContractOptions);
+    const account = this.getAccount();
 
-    const { contractId, transactionRequest } = this.createTransactionRequest(deployContractOptions);
-
-    const txCost = await this.account.provider.getTransactionCost(transactionRequest);
+    await account.sendTransaction(transactionRequest, {
+      awaitExecution: true,
+    });
 
-    const { maxFee: setMaxFee } = deployContractOptions;
+    return new Contract(contractId, this.interface, account);
+  }
 
-    if (isDefined(setMaxFee)) {
-      if (txCost.maxFee.gt(setMaxFee)) {
-        throw new FuelError(
-          ErrorCode.MAX_FEE_TOO_LOW,
-          `Max fee '${deployContractOptions.maxFee}' is lower than the required: '${txCost.maxFee}'.`
-        );
-      }
-    } else {
-      transactionRequest.maxFee = txCost.maxFee;
-    }
+  /**
+   * Deploys a contract asynchronously.
+   *
+   * @param deployContractOptions - The options for deploying the contract (optional).
+   * @returns An object containing the deployed contract and the transaction response.
+   */
+  async deployContractAsync(deployContractOptions: DeployContractOptions = {}) {
```

Comment:
There is a third option, which is to use TypeScript generics to infer if `awaitExecution` was specified, and then make `deployContract` have two possible return types. This was suggested by @nedsalk:

```ts
async deployContractNedim<
  AwaitExecution extends boolean = false,
  R = AwaitExecution extends true
    ? Contract
    : { transactionResponse: TransactionResponse; contract: Contract },
>(
  deployContractOptions: DeployContractOptions = {},
  awaitExecution?: AwaitExecution
): Promise<R> {
  if (!this.account) {
    throw new FuelError(ErrorCode.ACCOUNT_REQUIRED, 'Cannot deploy Contract without an account.');
  }
  const { configurableConstants } = deployContractOptions;
  if (configurableConstants) {
    this.setConfigurableConstants(configurableConstants);
  }
  const { contractId, transactionRequest } = this.createTransactionRequest(deployContractOptions);
  const txCost = await this.account.provider.getTransactionCost(transactionRequest);
  const { maxFee: setMaxFee } = deployContractOptions;
  if (isDefined(setMaxFee)) {
    if (txCost.maxFee.gt(setMaxFee)) {
      throw new FuelError(
        ErrorCode.MAX_FEE_TOO_LOW,
        `Max fee '${deployContractOptions?.maxFee}' is lower than the required: '${txCost.maxFee}'.`
      );
    }
  } else {
    transactionRequest.maxFee = txCost.maxFee;
  }
  await this.account.fund(transactionRequest, txCost);
  const transactionResponse = await this.account.sendTransaction(transactionRequest, {
    awaitExecution,
  });

  const contract = new Contract(contractId, this.interface, this.account);

  return (awaitExecution === true ? contract : { contract, transactionResponse }) as R;
}
```

Reply:
The reasoning behind this is
- we don't need to duplicate logic in a second `deployContractAsync` method when they're both doing the same thing,
- `awaitExecution` already exists as a parameter in `Provider.sendTransaction` and `Account.sendTransaction` so we'd be following the same pattern

Besides this, regardless of the approach taken, we need to apply it to typegen as well. Do we want to do it in this PR or in a subsequent one?